File: java/src/main/java/datastructures/Graph.java
package datastructures;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Queue;
import java.util.Set;

public class Graph {

    private final Map<Integer, List<Integer>> adjList = new HashMap<>();

    public void addVertex(int v) {
        adjList.putIfAbsent(v, new ArrayList<>());
    }

    public void addEdge(int u, int v) {
        adjList.putIfAbsent(u, new ArrayList<>());
        adjList.putIfAbsent(v, new ArrayList<>());
        adjList.get(u).add(v);
    }

    /**
     * BFS - level-order traversal from start node.
     * Use when: shortest path in unweighted graph, level-by-level processing.
     */
    public List<Integer> bfs(int start) {
        List<Integer> visited = new ArrayList<>();
        Set<Integer> seen = new HashSet<>();
        Queue<Integer> queue = new ArrayDeque<>();

        queue.offer(start);
        seen.add(start);

        while (!queue.isEmpty()) {
            int node = queue.poll();
            visited.add(node);
            for (int neighbour : adjList.getOrDefault(node, new ArrayList<>())) {
                if (seen.add(neighbour)) {
                    queue.offer(neighbour);
                }
            }
        }
        return visited;
    }

    /**
     * DFS - depth-first traversal from start node (iterative).
     * Use when: cycle detection, connected components, topological sort.
     */
    public List<Integer> dfs(int start) {
        List<Integer> visited = new ArrayList<>();
        Set<Integer> seen = new HashSet<>();
        Deque<Integer> stack = new ArrayDeque<>();

        stack.push(start);

        while (!stack.isEmpty()) {
            int node = stack.pop();
            if (!seen.add(node)) {
                continue;
            }
            visited.add(node);
            for (int neighbour : adjList.getOrDefault(node, new ArrayList<>())) {
                if (!seen.contains(neighbour)) {
                    stack.push(neighbour);
                }
            }
        }
        return visited;
    }
}

File: python/datastructures/graph.py
from collections import deque


class Graph:
    def __init__(self):
        self.adj_list = {}

    def add_vertex(self, v):
        self.adj_list.setdefault(v, [])

    def add_edge(self, u, v):
        self.adj_list.setdefault(u, [])
        self.adj_list.setdefault(v, [])
        self.adj_list[u].append(v)

    def bfs(self, start):
        visited = []
        seen = {start}
        queue = deque([start])

        while queue:
            node = queue.popleft()
            visited.append(node)
            for neighbour in self.adj_list.get(node, []):
                if neighbour not in seen:
                    seen.add(neighbour)
                    queue.append(neighbour)

        return visited

    def dfs(self, start):
        visited = []
        seen = set()
        stack = [start]

        while stack:
            node = stack.pop()
            if node in seen:
                continue
            seen.add(node)
            visited.append(node)
            for neighbour in self.adj_list.get(node, []):
                if neighbour not in seen:
                    stack.append(neighbour)

        return visited

